### ISSUE TITLE
SameDiff: remove outputs() variable inference in favor of user specified outputs

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#8265]
The SameDiff.outputs() method is a "best guess" inference of the predictions of the SameDiff instance.

https://github.com/eclipse/deeplearning4j/blob/master/nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java#L1824-L1831

This inference is unfortunately unreliable: the "best guess" can fail (and confuse users), for example when a loss function is added.
Even if we filter loss variables, we still can't infer automatically in the case of multi-op loss functions (like softmax -> difference -> squared difference -> (loss = sum(squared difference)).

Instead:
* User should call setOutputs(String... ) to set the outputs explicitly
* outputs() should throw a useful exception if outputs have not been set
* We should have no-arg feed forward methods that use these stored output variable names (or again, exception if not set)

And of course, outputs should be serialized with the graph.
